python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -846,38 +851,77 @@ def get_updated_parent_asset_partitions(
     ) -> AbstractSet[AssetKey]:
         """Return the set of root unreconciled ancestors of the given asset partition, i.e. the set
         of ancestors of this asset partition whose parents have been updated more recently than
         they have.
         """
-        if self.asset_graph.is_source(asset_partition.asset_key):
-            return set()
+        if asset_partition in self._root_unreconciled_ancestors_cache:
+            return self._root_unreconciled_ancestors_cache[asset_partition]
 
-        parent_asset_partitions = self.asset_graph.get_parents_partitions(
-            dynamic_partitions_store=self,
-            current_time=self._evaluation_time,
-            asset_key=asset_partition.asset_key,
-            partition_key=asset_partition.partition_key,
-        ).parent_partitions
-
-        updated_parents = self.get_updated_parent_asset_partitions(
-            asset_partition,
-            parent_asset_partitions,
-            respect_materialization_data_versions=respect_materialization_data_versions,
-        )
+        # First gather all the candidates that weren't previously computed as well as their parents
+        visited: set[AssetKeyPartitionKey] = set()
 
-        root_unreconciled_ancestors = {asset_partition.asset_key} if updated_parents else set()
+        queue: deque[AssetKeyPartitionKey] = deque()
+        queue.append(asset_partition)
 
-        # recurse over parents
-        for parent in set(parent_asset_partitions) - updated_parents:
-            root_unreconciled_ancestors.update(
-                self.get_root_unreconciled_ancestors(
-                    asset_partition=parent,
-                    respect_materialization_data_versions=respect_materialization_data_versions,
+        while queue:
+            current_partition = queue.popleft()
+            visited.add(current_partition)
+
+            if self.asset_graph.is_source(current_partition.asset_key):
+                continue
+
+            parent_asset_partitions = self.asset_graph.get_parents_partitions(
+                dynamic_partitions_store=self,
+                current_time=self._evaluation_time,
+                asset_key=current_partition.asset_key,
+                partition_key=current_partition.partition_key,
+            ).parent_partitions
+
+            for parent in parent_asset_partitions:
+                if parent not in visited and parent not in self._root_unreconciled_ancestors_cache:
+                    queue.append(parent)
+
+        # Toposort them so that at each iteration we can count on the cache being full for
+        # all of your parents, then update the cache for each node based on the parent's results
+        toposort_queue = ToposortedPriorityQueue(self.asset_graph, visited)
+
+        while len(toposort_queue) > 0:
+            candidates_unit = toposort_queue.dequeue()
+            for current_partition in candidates_unit:
+                if self.asset_graph.is_source(current_partition.asset_key):
```

Comment:
Have we not already filtered out the sources?

Reply:
updated the logic here to short-circuit earlier if its a source asset instead of bothering to put it in the cache, since we know the result right away